icelake/src/io/task_writer.rs

```diff
@@ -147,3 +169,216 @@ impl UnpartitionedWriter {
+    async fn split_by_partition(
+        &mut self,
+        batch: &RecordBatch,
+    ) -> Result<HashMap<OwnedRow, RecordBatch>> {
+        let value_array = Arc::new(StructArray::from(
+            self.partition_spec
+                .fields
+                .iter()
+                .map(|field| {
+                    let array: ArrayRef = batch
+                        .column_by_name(&field.name)
+                        .ok_or_else(|| {
+                            crate::error::Error::new(
+                                crate::ErrorKind::IcebergDataInvalid,
+                                format!(
+                                    "Can't find column according partition spec, column name {}",
+                                    field.name
+                                ),
+                            )
+                        })?
+                        .clone();
+                    let transform_func = create_transform_function(&field.transform)?;
+                    let array = transform_func.transform(array)?;
+                    let field = Arc::new(Field::new(
+                        field.name.clone(),
+                        array.data_type().clone(),
+                        true,
+                    ));
+                    Ok((field, array))
+                })
+                .collect::<Result<Vec<_>>>()?,
+        ));
+
+        let mut row_converter = RowConverter::new(vec![SortField::new(
+            value_array.data_type().clone(),
+        )])
+        .map_err(|e| crate::error::Error::new(crate::ErrorKind::ArrowError, format!("{}", e)))?;
+        let rows = row_converter
+            .convert_columns(&[value_array.clone()])
+            .map_err(|e| {
+                crate::error::Error::new(crate::ErrorKind::ArrowError, format!("{}", e))
+            })?;
+
+        // Group the batch by row value.
+        let mut group_ids = HashMap::new();
+        rows.into_iter().enumerate().for_each(|(row_id, row)| {
+            group_ids.entry(row.owned()).or_insert(vec![]).push(row_id);
+        });
+
+        // Create the partition group if not exist.
+        for (row, row_ids) in group_ids.iter() {
+            let row_id = row_ids[0];
+            if let Entry::Vacant(entry) = self.groups.entry(row.clone()) {
+                let partition_array = value_array.slice(row_id, 1);
+                let writer = DataFileWriter::try_new(
+                    self.operator.clone(),
+                    self.table_location.clone(),
+                    self.location_generator.clone(),
+                    self.schema.clone(),
+                    1024,
+                    1024 * 1024,
+                    self.table_config.clone(),
+                )
+                .await?;
+                entry.insert(PartitionGroup {
+                    partition_array,
+                    writer,
+                });
+            }
+        }
```

Comment:
I don't quite understand why we need to do the insertion here?

Reply:
For batch belong to different partition value, it need be wrote using different writer. And we store this writer in 
`HashMap<OwnedRow,ParitionGroup>`
So every time we meet a partition value we never see it before, we should create a new writer(PartitionGroup) for it.

And why we need StructArray? After we close writer, we need to write partition value into DataFile. We store this partition value in this StructArray. Actually this StructArray only have one element. And when we close writer, we convert this StructArrary into `Vec<AnyValue>` and store it in DataFile.

Reply:
Oh, I see. Please add these as comments in the field.